File: nlm_controller/Demo/CORTEX_M4_NXP_LA9310_GCC/ps_nr_pbch_bit_deinterleaver.c

#include "FreeRTOS.h"
#include "task.h"
#include "debug_console.h"
#include "pbch.h"
#include <string.h>

uint32_t ps_nr_pbch_bit_deinterleaver(uint32_t bitIn){
    
    uint32_t bitOut = 0; 
    uint8_t i = 0;
    
    uint8_t pbchInterleaverPattern[32] = {16,23,18,17,8,30,10,6,24,7,0,5,3,2,1,4,9,11,12,13,14,15,19,20,21,22,25,26,27,28,29,31};
    
    // First 6 SFN Bits
    uint8_t jsfn    = 0;
    uint8_t jother  = 14;
    uint8_t jhrf    = 10;
    uint8_t jssb    = 11;
    
    bitOut |= (bitIn>>pbchInterleaverPattern[jother])&0x1;
    i = i+1;
    jother++;

    for(i=1;i<7;i++){
      bitOut |= ((bitIn>>pbchInterleaverPattern[jsfn])&0x1)<<i;
      jsfn++;
    }

    for(i=7;i<24;i++){
      bitOut |= ((bitIn>>pbchInterleaverPattern[jother])&0x1)<<i;
      jother++;
    }
    
    for(i=24;i<28;i++){
      bitOut |= ((bitIn>>pbchInterleaverPattern[jsfn])&0x1)<<i;
      jsfn++;
    }

    bitOut |= ((bitIn>>pbchInterleaverPattern[jhrf])&0x1)<<i;

    for(i=29;i<32;i++){
      bitOut |= ((bitIn>>pbchInterleaverPattern[jssb])&0x1)<<i;
      jssb++;
    }
    
    return bitOut;
}

File: nlm_floating_point/PBCH/ps_nr_pbch_bit_deinterleaver.m
function bitOut  = ps_nr_pbch_bit_deinterleaver(bitIn)
  
  pbchPayloadSize = 32;
  jsfn = 0;
  jhrf = 10;
  jssb = 11;
  jother = 14;

  pbchPayloadInterleaverPattern = [16,23,18,17,8,30,10,6,24,7,0,5,3,2,1,4,9,11,12,13,14,15,19,20,21,22,25,26,27,28,29,31];

  %pbchPayloadInterleaverPattern = [10,14,13,12,15,11,7,9,4,16,6,17,18,19,20,21,0,3,2,22,23,24,25,1,8,22,26,27,27,29,5,31];

  bitOut = [];
  
  % First 6 SFN Bits
  jsfn    = 0;
  jother  = 14;
  jhrf    = 10;
  jssb    = 11;

  k = 1;
  bitOut(k) = bitIn(pbchPayloadInterleaverPattern(jother+1) + 1);
  k = k+1;
  jother = jother+1;
  
  for i = 1:1:6
      bitOut(k) = bitIn(pbchPayloadInterleaverPattern(jsfn+1) + 1);
      k = k+1;
      jsfn = jsfn+1;
  end    
  
  for i = 1:1:17
      bitOut(k) = bitIn(pbchPayloadInterleaverPattern(jother+1) + 1);
      k = k+1;
      jother = jother+1;
  end   

  for i = 1:1:4
      bitOut(k) = bitIn(pbchPayloadInterleaverPattern(jsfn+1) + 1);
      k = k+1;
      jsfn = jsfn+1;
  end
  
  bitOut(k) = bitIn(pbchPayloadInterleaverPattern(jhrf+1) + 1);
  k = k+1;
  
  for i = 1:1:3
      bitOut(k) = bitIn(pbchPayloadInterleaverPattern(jssb+1) + 1);
      k = k+1;
      jssb = jssb+1;
  end
 
end
